File: src/hillclimber/evaluator.ts
/**
 * HillClimber Evaluator Module
 *
 * Real-time progress scoring during task execution.
 * Runs verification, parses test output, returns structured feedback.
 *
 * Part of the MAP-inspired architecture for 10x better HillClimber.
 */

import { Effect } from "effect";
import { existsSync } from "node:fs";
import { join } from "node:path";
import type { TerminalBenchTask } from "../bench/terminal-bench.js";
import { runTB2InDocker } from "../bench/tb2-docker-runner.js";

// ============================================================================
// Types
// ============================================================================

export interface FailureDetail {
  /** Test case identifier (e.g., "test_regex_matches_dates") */
  testName: string;
  /** Line number in test file, if available */
  lineNumber?: number;
  /** What was expected */
  expected?: string;
  /** What was actually received */
  actual?: string;
  /** Raw error message */
  message: string;
}

export interface EvaluatorResult {
  /** Overall pass/fail status */
  passed: boolean;
  /** Progress score 0-1 (tests passing / total tests) */
  progress: number;
  /** Total number of tests */
  testsTotal: number;
  /** Number of tests passing */
  testsPassing: number;
  /** Detailed failure information */
  failures: FailureDetail[];
  /** Suggestion for what to try next */
  suggestion?: string;
  /** Raw verification output */
  rawOutput: string;
  /** Duration in ms */
  durationMs: number;
}

// ============================================================================
// Verification Runner
// ============================================================================

/**
 * Run verification command and return raw output.
 */
export const runVerification = (
  task: TerminalBenchTask,
  workspace: string
): Effect.Effect<{ exitCode: number; output: string }, Error> =>
  Effect.tryPromise({
    try: async () => {
      const startTime = Date.now();
      const cmd = task.verification.command ?? task.verification.script ?? "exit 1";

      // Change to workspace directory
      const originalCwd = process.cwd();
      process.chdir(workspace);

      try {
        const proc = Bun.spawn(["sh", "-c", cmd], {
          stdout: "pipe",
          stderr: "pipe",
          cwd: workspace,
          env: {
            ...process.env,
            PYTHONUNBUFFERED: "1",
          },
        });

        const exitCode = await proc.exited;
        const stdout = await new Response(proc.stdout).text();
        const stderr = await new Response(proc.stderr).text();

        const durationMs = Date.now() - startTime;

        return {
          exitCode,
          output: stdout + (stderr ? `\nSTDERR:\n${stderr}` : ""),
          durationMs,
        };
      } finally {
        process.chdir(originalCwd);
      }
    },
    catch: (e) => new Error(`Verification failed: ${e}`),
  });

/**
 * Run verification using Docker for TB2 tasks.
 *
 * This is the proper way to run TB2 tests - in the Docker environment
 * where /app/ exists as expected. Uses blind verification (pass/fail only).
 *
 * @param task - TB2 task with source_path pointing to task directory
 * @param workspace - Local workspace with the solution
 */
export const runVerificationWithDocker = (
  task: TerminalBenchTask,
  workspace: string
): Effect.Effect<{ exitCode: number; output: string; passed: boolean; progress: number; testsPassing: number; testsTotal: number }, Error> =>
  Effect.tryPromise({
    try: async () => {
      // Check if task has source_path with tests
      const taskDir = task.source_path;
      if (!taskDir || !existsSync(taskDir)) {
        throw new Error(`Task source_path not found: ${taskDir}. Cannot run Docker verification.`);
      }

      const testsDir = join(taskDir, "tests");
      if (!existsSync(testsDir)) {
        throw new Error(`Tests directory not found: ${testsDir}. Cannot run Docker verification.`);
      }

      // Run verification in Docker
      const result = await runTB2InDocker({
        taskId: task.id,
        taskDir,
        workspace,
        timeout: 120000,
        captureDetails: false, // Blind verification - no expected values
      });

      // Format output for compatibility with existing parsers
      const output = result.passed
        ? `${result.testsPassing} passed in ${result.durationMs}ms`
        : `${result.testsPassing} passed, ${result.testsTotal - result.testsPassing} failed in ${result.durationMs}ms${result.feedback ? `\n${result.feedback}` : ""}`;

      return {
        exitCode: result.exitCode,
        output,
        passed: result.passed,
        progress: result.progress,
        testsPassing: result.testsPassing,
        testsTotal: result.testsTotal,
      };
    },
    catch: (e) => new Error(`Docker verification failed: ${e}`),
  });

// ============================================================================
// Output Parsers
// ============================================================================

/**
 * Parse pytest output to extract test results.
 */
export function parsePytestOutput(output: string): {
  total: number;
  passed: number;
  failed: number;
  failures: FailureDetail[];
} {
  const failures: FailureDetail[] = [];

  // Parse pytest summary line (e.g., "1 passed, 2 failed in 0.05s")
  const summaryMatch = output.match(/(\d+)\s+passed.*?(\d+)\s+failed/i);
  const passedOnlyMatch = output.match(/(\d+)\s+passed/i);
  const failedOnlyMatch = output.match(/(\d+)\s+failed/i);

  let passed = 0;
  let failed = 0;

  if (summaryMatch) {
    passed = parseInt(summaryMatch[1], 10);
    failed = parseInt(summaryMatch[2], 10);
  } else if (passedOnlyMatch && !failedOnlyMatch) {
    passed = parseInt(passedOnlyMatch[1], 10);
    failed = 0;
  } else if (failedOnlyMatch) {
    failed = parseInt(failedOnlyMatch[1], 10);
    passed = passedOnlyMatch ? parseInt(passedOnlyMatch[1], 10) : 0;
  }

  // Parse individual test failures
  // Pattern: FAILED tests/test_*.py::test_name - AssertionError: message
  const failurePattern = /FAILED\s+(\S+)::(\w+)\s*[-–]\s*(.+?)(?=\n(?:FAILED|PASSED|===|$))/gs;
  let match;

  while ((match = failurePattern.exec(output)) !== null) {
    failures.push({
      testName: match[2],
      message: match[3].trim(),
    });
  }

  // Also parse assertion details from verbose output
  // Pattern: AssertionError: Expected [...], but got [...]
  const assertPattern = /AssertionError:\s*Expected\s+(.+?),\s+but\s+got\s+(.+?)(?=\n|$)/gi;

  while ((match = assertPattern.exec(output)) !== null) {
    // Find the corresponding failure and add details
    if (failures.length > 0) {
      const lastFailure = failures[failures.length - 1];
      if (!lastFailure.expected) {
        lastFailure.expected = match[1].trim();
        lastFailure.actual = match[2].trim();
      }
    }
  }

  // Parse specific pattern for regex-log task
  // "Expected ['date1', 'date2'], but got ['date1', 'date3']"
  const listAssertPattern = /Expected\s+(\[.+?\]),\s+but\s+got\s+(\[.+?\])/gi;

  while ((match = listAssertPattern.exec(output)) !== null) {
    if (failures.length > 0) {
      const lastFailure = failures[failures.length - 1];
      if (!lastFailure.expected) {
        lastFailure.expected = match[1];
        lastFailure.actual = match[2];
      }
    }
  }

  const total = passed + failed;

  return { total, passed, failed, failures };
}

/**
 * Parse generic test output (exit code based).
 */
export function parseGenericOutput(output: string, exitCode: number): {
  total: number;
  passed: number;
  failed: number;
  failures: FailureDetail[];
} {
  // Simple: if exit code is 0, all tests passed
  if (exitCode === 0) {
    return { total: 1, passed: 1, failed: 0, failures: [] };
  }

  // Try to extract error information from output
  const failures: FailureDetail[] = [];

  // Look for common error patterns
  const errorPatterns = [
    /error:\s*(.+)/gi,
    /failed:\s*(.+)/gi,
    /assertion.*?:\s*(.+)/gi,
    /expected\s+(.+?)\s+but\s+got\s+(.+)/gi,
  ];

  for (const pattern of errorPatterns) {
    let match;
    while ((match = pattern.exec(output)) !== null) {
      failures.push({
        testName: "verification",
        message: match[0],
        expected: match[1],
        actual: match[2],
      });
    }
  }

  // If no failures parsed, create a generic one
  if (failures.length === 0) {
    failures.push({
      testName: "verification",
      message: output.slice(0, 500) + (output.length > 500 ? "..." : ""),
    });
  }

  return { total: 1, passed: 0, failed: 1, failures };
}

// ============================================================================
// Suggestion Generator
// ============================================================================

/**
 * Generate a suggestion based on failures.
 */
export function generateSuggestion(
  taskId: string,
  failures: FailureDetail[]
): string | undefined {
  if (failures.length === 0) {
    return undefined;
  }

  // Task-specific suggestions
  if (taskId === "regex-log") {
    const firstFailure = failures[0];
    if (firstFailure.expected && firstFailure.actual) {
      // Analyze the difference (with error handling)
      let expected: unknown[] = [];
      let actual: unknown[] = [];
      try {
        expected = JSON.parse(firstFailure.expected || "[]");
        actual = JSON.parse(firstFailure.actual || "[]");
      } catch (e) {
        // If JSON parsing fails, try to extract arrays from string representation
        const expectedMatch = firstFailure.expected.match(/\[(.*?)\]/);
        const actualMatch = firstFailure.actual.match(/\[(.*?)\]/);
        if (expectedMatch) {
          expected = expectedMatch[1].split(",").map(s => s.trim().replace(/['"]/g, ""));
        }
        if (actualMatch) {
          actual = actualMatch[1].split(",").map(s => s.trim().replace(/['"]/g, ""));
        }
      }

      if (expected.length > actual.length) {
        return `Missing ${expected.length - actual.length} matches. Check if regex is too restrictive.`;
      } else if (actual.length > expected.length) {
        return `${actual.length - expected.length} false positives. Check boundary conditions and IP validation.`;
      } else {
        // Same length but different values
        const wrongMatches = actual.filter((a: string) => !expected.includes(a));
        if (wrongMatches.length > 0) {
          return `Wrong matches: ${wrongMatches.join(", ")}. Check date validation or IP boundary conditions.`;
        }
      }
    }
    return "Check regex for: 1) Valid IPv4 (0-255 per octet), 2) Valid dates (month 1-12, correct days), 3) Boundary assertions.";
  }

  if (taskId === "path-tracing") {
    return "Check: 1) PPM format (P6 header), 2) Image dimensions match reference, 3) Path tracing algorithm correctness.";
  }

  if (taskId === "model-extraction-relu-logits") {
    return "Check: 1) Query strategy (try unit vectors), 2) ReLU activation handling, 3) Row matching tolerance.";
  }

  if (taskId === "video-processing") {
    return "Check: 1) Frame difference calculation, 2) Motion threshold tuning, 3) TOML output format.";
  }

  if (taskId === "dna-assembly") {
    return "Check: 1) BsaI site structure, 2) Melting temperature calculation, 3) Overhang uniqueness.";
  }

  // Generic suggestion
  return `Fix: ${failures[0].message.slice(0, 100)}`;
}

// ============================================================================
// Main Evaluator
// ============================================================================

/**
 * Evaluate progress on a task by running verification and parsing results.
 *
 * @param task Terminal-Bench task definition
 * @param workspace Working directory containing task files
 * @returns Structured evaluation result with progress score and failures
 */
export const evaluateProgress = (
  task: TerminalBenchTask,
  workspace: string
): Effect.Effect<EvaluatorResult, Error, never> =>
  Effect.gen(function* () {
    const startTime = Date.now();

    // Run verification
    const { exitCode, output } = yield* runVerification(task, workspace);

    const durationMs = Date.now() - startTime;

    // Parse output based on verification type
    let parseResult: { total: number; passed: number; failed: number; failures: FailureDetail[] };

    if (task.verification.type === "test" || output.includes("pytest") || output.includes("PASSED") || output.includes("FAILED")) {
      parseResult = parsePytestOutput(output);
    } else {
      parseResult = parseGenericOutput(output, exitCode);
    }

    // Calculate progress
    const progress = parseResult.total > 0 ? parseResult.passed / parseResult.total : 0;
    const passed = exitCode === 0;

    // Generate suggestion
    const suggestion = generateSuggestion(task.id, parseResult.failures);

    const result: EvaluatorResult = {
      passed,
      progress,
      testsTotal: parseResult.total,
      testsPassing: parseResult.passed,
      failures: parseResult.failures,
      rawOutput: output,
      durationMs,
    };

    if (suggestion !== undefined) {
      result.suggestion = suggestion;
    }

    return result;
  });

/**
 * Evaluate progress using Docker-based verification for TB2 tasks.
 *
 * This uses the proper Docker environment where /app/ exists.
 * Returns blind verification results (pass/fail + progress, no expected values).
 *
 * @param task Terminal-Bench task definition (must have source_path)
 * @param workspace Working directory containing solution files
 * @returns Structured evaluation result with progress score
 */
export const evaluateProgressWithDocker = (
  task: TerminalBenchTask,
  workspace: string
): Effect.Effect<EvaluatorResult, Error, never> =>
  Effect.gen(function* () {
    const result = yield* runVerificationWithDocker(task, workspace);

    // For blind verification, we don't parse individual failures
    // We only know pass/fail and generic progress
    const evalResult: EvaluatorResult = {
      passed: result.passed,
      progress: result.progress,
      testsTotal: result.testsTotal,
      testsPassing: result.testsPassing,
      failures: [], // Blind - no individual failure details
      rawOutput: result.output,
      durationMs: 0, // Not tracked in Docker result
    };

    // Add generic suggestion for failures (no expected values)
    if (!result.passed && result.testsTotal > 0) {
      const failCount = result.testsTotal - result.testsPassing;
      evalResult.suggestion = `${failCount} test${failCount > 1 ? "s" : ""} failing. Review edge cases and constraints.`;
    }

    return evalResult;
  });

/**
 * Quick evaluation that doesn't parse detailed failures.
 * Use for frequent checks during execution.
 */
export const quickEvaluate = async (
  task: TerminalBenchTask,
  workspace: string
): Promise<{ passed: boolean; progress: number; message: string }> => {
  try {
    const cmd = task.verification.command ?? task.verification.script ?? "exit 1";

    const originalCwd = process.cwd();
    process.chdir(workspace);

    try {
      const proc = Bun.spawn(["sh", "-c", cmd], {
        stdout: "pipe",
        stderr: "pipe",
        cwd: workspace,
      });

      const exitCode = await proc.exited;
      const stdout = await new Response(proc.stdout).text();

      // Quick parse for pass count
      const passedMatch = stdout.match(/(\d+)\s+passed/i);
      const failedMatch = stdout.match(/(\d+)\s+failed/i);

      const passed = passedMatch ? parseInt(passedMatch[1], 10) : 0;
      const failed = failedMatch ? parseInt(failedMatch[1], 10) : 0;
      const total = passed + failed || 1;

      const progress = exitCode === 0 ? 1 : passed / total;

      return {
        passed: exitCode === 0,
        progress,
        message: exitCode === 0 ? "All tests passing" : `${passed}/${total} tests passing`,
      };
    } finally {
      process.chdir(originalCwd);
    }
  } catch (e) {
    return {
      passed: false,
      progress: 0,
      message: `Evaluation error: ${e}`,
    };
  }
};

/**
 * Format evaluation result for injection into FM prompt.
 */
export function formatForPrompt(result: EvaluatorResult): string {
  const lines: string[] = [];

  if (result.passed) {
    lines.push(`Verification: PASSED (${result.testsPassing}/${result.testsTotal} tests)`);
  } else {
    lines.push(`Verification: FAILED (${result.testsPassing}/${result.testsTotal} tests)`);

    // Add first 3 failures
    for (const failure of result.failures.slice(0, 3)) {
      if (failure.expected && failure.actual) {
        lines.push(`  - ${failure.testName}: expected ${failure.expected}, got ${failure.actual}`);
      } else {
        lines.push(`  - ${failure.testName}: ${failure.message.slice(0, 100)}`);
      }
    }

    if (result.failures.length > 3) {
      lines.push(`  ... and ${result.failures.length - 3} more failures`);
    }

    if (result.suggestion) {
      lines.push(`Suggestion: ${result.suggestion}`);
    }
  }

  return lines.join("\n");
}

File: crates/hillclimber/src/evaluator.rs
//! HillClimber Evaluator Module
//!
//! Real-time progress scoring during task execution.
//! Runs verification, parses test output, returns structured feedback.
//!
//! Supports Docker-based verification via bollard and local pytest fallback.

use crate::error::{HillClimberError, Result};
use crate::types::{EvaluatorResult, FailureDetail, TerminalBenchTask};
use bollard::container::{Config, CreateContainerOptions, LogOutput, LogsOptions, RemoveContainerOptions, WaitContainerOptions};
use bollard::image::CreateImageOptions;
use bollard::Docker;
use futures::StreamExt;
use regex::Regex;
use std::path::Path;
use std::time::Instant;
use tokio::process::Command;

// ============================================================================
// Docker Image
// ============================================================================

const TB2_IMAGE: &str = "terminal-bench-2:latest";

// ============================================================================
// Pytest Output Parsing
// ============================================================================

/// Parse pytest output to extract test results.
pub fn parse_pytest_output(output: &str) -> ParseResult {
    let mut failures: Vec<FailureDetail> = Vec::new();

    // Parse pytest summary line - handles both orders:
    // "1 passed, 2 failed in 0.05s" OR "2 failed, 1 passed in 0.05s"
    let passed_re = Regex::new(r"(\d+)\s+passed").unwrap();
    let failed_re = Regex::new(r"(\d+)\s+failed").unwrap();
    // Also parse our own summary format: "Verification: FAILED (10/21 tests)"
    let summary_format_re = Regex::new(r"Verification:\s*\w+\s*\((\d+)/(\d+)\s*tests?\)").unwrap();

    let (mut passed, mut failed) = (0u32, 0u32);

    // First try our summary format (takes precedence since it's already parsed)
    if let Some(caps) = summary_format_re.captures(output) {
        passed = caps[1].parse().unwrap_or(0);
        let total: u32 = caps[2].parse().unwrap_or(0);
        failed = total.saturating_sub(passed);
    } else {
        // Parse pytest format: passed and failed independently (order doesn't matter)
        if let Some(caps) = passed_re.captures(output) {
            passed = caps[1].parse().unwrap_or(0);
        }
        if let Some(caps) = failed_re.captures(output) {
            failed = caps[1].parse().unwrap_or(0);
        }
    }

    // Parse individual test failures from short summary
    // Pattern: FAILED tests/test_*.py::test_name
    let short_failure_re = Regex::new(r"FAILED\s+(\S+)::(\w+)").unwrap();

    // Pattern: tests/test_*.py::test_name FAILED - message (verbose output)
    let verbose_failure_re = Regex::new(r"(\S+)::(\w+)\s+FAILED\s*[-–]\s*(.+?)(?:\n|$)").unwrap();

    // First try to parse verbose output with messages
    for caps in verbose_failure_re.captures_iter(output) {
        failures.push(FailureDetail {
            test_name: caps[2].to_string(),
            line_number: None,
            expected: None,
            actual: None,
            message: caps[3].trim().to_string(),
        });
    }

    // If no failures found, try short summary format
    if failures.is_empty() {
        for caps in short_failure_re.captures_iter(output) {
            failures.push(FailureDetail {
                test_name: caps[2].to_string(),
                line_number: None,
                expected: None,
                actual: None,
                message: "Test failed".to_string(),
            });
        }
    }

    // Also parse assertion details from verbose output
    // Pattern: AssertionError: Expected [...], but got [...]
    let assert_re = Regex::new(r"Expected\s+(\[.+?\]),\s+but\s+got\s+(\[.+?\])").unwrap();

    for caps in assert_re.captures_iter(output) {
        if let Some(last) = failures.last_mut() {
            if last.expected.is_none() {
                last.expected = Some(caps[1].to_string());
                last.actual = Some(caps[2].to_string());
            }
        }
    }

    let total = passed + failed;

    ParseResult {
        total,
        passed,
        failed,
        failures,
    }
}

/// Parse generic test output (exit code based).
pub fn parse_generic_output(output: &str, exit_code: i32) -> ParseResult {
    if exit_code == 0 {
        return ParseResult {
            total: 1,
            passed: 1,
            failed: 0,
            failures: vec![],
        };
    }

    // Try to extract error information from output
    let mut failures: Vec<FailureDetail> = Vec::new();

    // Look for common error patterns
    let error_patterns = [
        r"error:\s*(.+)",
        r"failed:\s*(.+)",
        r"assertion.*?:\s*(.+)",
    ];

    for pattern in error_patterns {
        if let Ok(re) = Regex::new(pattern) {
            for caps in re.captures_iter(output) {
                failures.push(FailureDetail {
                    test_name: "verification".to_string(),
                    line_number: None,
                    expected: None,
                    actual: None,
                    message: caps[0].to_string(),
                });
            }
        }
    }

    // If no failures parsed, create a generic one
    if failures.is_empty() {
        let truncated = if output.len() > 500 {
            format!("{}...", &output[..500])
        } else {
            output.to_string()
        };
        failures.push(FailureDetail {
            test_name: "verification".to_string(),
            line_number: None,
            expected: None,
            actual: None,
            message: truncated,
        });
    }

    ParseResult {
        total: 1,
        passed: 0,
        failed: 1,
        failures,
    }
}

/// Parsed test results.
pub struct ParseResult {
    pub total: u32,
    pub passed: u32,
    pub failed: u32,
    pub failures: Vec<FailureDetail>,
}

// ============================================================================
// Local Pytest Runner
// ============================================================================

/// Run pytest locally (fallback when Docker unavailable).
pub async fn run_local_pytest(workspace: &Path) -> Result<EvaluatorResult> {
    let start = Instant::now();

    let tests_dir = workspace.join("tests");
    if !tests_dir.exists() {
        return Err(HillClimberError::Workspace(format!(
            "Tests directory not found: {}",
            tests_dir.display()
        )));
    }

    let output = Command::new("python3")
        .args(["-m", "pytest", "tests/", "-v", "--tb=short"])
        .current_dir(workspace)
        .env("PYTHONDONTWRITEBYTECODE", "1")
        .output()
        .await
        .map_err(|e| HillClimberError::Workspace(format!("Failed to run pytest: {}", e)))?;

    let duration_ms = start.elapsed().as_millis() as u64;

    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);
    let raw_output = format!("{}\n{}", stdout, stderr);

    let parse_result = parse_pytest_output(&raw_output);
    let passed = output.status.success() && parse_result.failed == 0;
    let progress = if parse_result.total > 0 {
        parse_result.passed as f64 / parse_result.total as f64
    } else {
        0.0
    };

    Ok(EvaluatorResult {
        passed,
        progress,
        tests_total: parse_result.total,
        tests_passing: parse_result.passed,
        failures: parse_result.failures,
        suggestion: None,
        raw_output,
        duration_ms,
    })
}

// ============================================================================
// Docker Verification
// ============================================================================

/// Check if Docker is available and daemon is running.
pub async fn is_docker_available() -> bool {
    // Check environment override
    if std::env::var("OPENAGENTS_DOCKER_AVAILABLE").as_deref() == Ok("0") {
        return false;
    }
    if std::env::var("OPENAGENTS_DOCKER_AVAILABLE").as_deref() == Ok("1") {
        return true;
    }

    match Docker::connect_with_local_defaults() {
        Ok(docker) => docker.ping().await.is_ok(),
        Err(_) => false,
    }
}

/// Run verification using Docker.
pub async fn run_docker_verification(
    task: &TerminalBenchTask,
    workspace: &Path,
    timeout_secs: u64,
) -> Result<EvaluatorResult> {
    let start = Instant::now();

    let docker = Docker::connect_with_local_defaults()
        .map_err(|e| HillClimberError::Docker(e))?;

    // Ensure image exists (pull if needed)
    let _ = docker
        .create_image(
            Some(CreateImageOptions {
                from_image: TB2_IMAGE,
                ..Default::default()
            }),
            None,
            None,
        )
        .collect::<Vec<_>>()
        .await;

    // Create container name
    let container_name = format!("hc-{}-{}", task.id, uuid::Uuid::new_v4().to_string()[..8].to_string());

    // Get absolute workspace path
    let workspace_abs = workspace
        .canonicalize()
        .map_err(|e| HillClimberError::Workspace(format!("Invalid workspace path: {}", e)))?;

    // Create container config
    let config = Config {
        image: Some(TB2_IMAGE.to_string()),
        cmd: Some(vec![
            "pytest".to_string(),
            "tests/".to_string(),
            "-v".to_string(),
            "--tb=short".to_string(),
        ]),
        working_dir: Some("/app".to_string()),
        host_config: Some(bollard::models::HostConfig {
            binds: Some(vec![format!(
                "{}:/app:rw",
                workspace_abs.display()
            )]),
            auto_remove: Some(true),
            ..Default::default()
        }),
        ..Default::default()
    };

    // Create container
    let container = docker
        .create_container(
            Some(CreateContainerOptions {
                name: &container_name,
                platform: None,
            }),
            config,
        )
        .await
        .map_err(|e| HillClimberError::Docker(e))?;

    // Start container
    docker
        .start_container::<String>(&container.id, None)
        .await
        .map_err(|e| HillClimberError::Docker(e))?;

    // Wait for container with timeout
    let wait_result = tokio::time::timeout(
        std::time::Duration::from_secs(timeout_secs),
        docker.wait_container(&container.id, Some(WaitContainerOptions { condition: "not-running" })).next(),
    )
    .await;

    let exit_code = match wait_result {
        Ok(Some(Ok(result))) => result.status_code,
        Ok(Some(Err(e))) => {
            // Try to remove container on error
            let _ = docker.remove_container(&container.id, Some(RemoveContainerOptions { force: true, ..Default::default() })).await;
            return Err(HillClimberError::Docker(e));
        }
        Ok(None) => {
            let _ = docker.remove_container(&container.id, Some(RemoveContainerOptions { force: true, ..Default::default() })).await;
            return Err(HillClimberError::Timeout("Docker container exited unexpectedly".to_string()));
        }
        Err(_) => {
            // Timeout - kill container
            let _ = docker.remove_container(&container.id, Some(RemoveContainerOptions { force: true, ..Default::default() })).await;
            return Err(HillClimberError::Timeout(format!(
                "Docker verification timed out after {}s",
                timeout_secs
            )));
        }
    };

    // Collect logs
    let logs_options = LogsOptions::<String> {
        stdout: true,
        stderr: true,
        follow: false,
        ..Default::default()
    };

    let logs: Vec<_> = docker
        .logs(&container.id, Some(logs_options))
        .collect()
        .await;

    let raw_output: String = logs
        .into_iter()
        .filter_map(|r| r.ok())
        .map(|log| match log {
            LogOutput::StdOut { message } | LogOutput::StdErr { message } => {
                String::from_utf8_lossy(&message).to_string()
            }
            _ => String::new(),
        })
        .collect::<Vec<_>>()
        .join("");

    let duration_ms = start.elapsed().as_millis() as u64;

    // Parse output
    let parse_result = if raw_output.contains("pytest") || raw_output.contains("PASSED") || raw_output.contains("FAILED") {
        parse_pytest_output(&raw_output)
    } else {
        parse_generic_output(&raw_output, exit_code as i32)
    };

    let passed = exit_code == 0;
    let progress = if parse_result.total > 0 {
        parse_result.passed as f64 / parse_result.total as f64
    } else if passed {
        1.0
    } else {
        0.0
    };

    Ok(EvaluatorResult {
        passed,
        progress,
        tests_total: parse_result.total,
        tests_passing: parse_result.passed,
        failures: parse_result.failures,
        suggestion: None,
        raw_output,
        duration_ms,
    })
}

// ============================================================================
// Main Evaluator Functions
// ============================================================================

/// Evaluate progress on a task by running verification.
///
/// Tries Docker first, falls back to local pytest if unavailable.
pub async fn evaluate_progress(
    task: &TerminalBenchTask,
    workspace: &Path,
) -> Result<EvaluatorResult> {
    // Check if Docker is available
    if is_docker_available().await {
        match run_docker_verification(task, workspace, 120).await {
            Ok(result) => return Ok(result),
            Err(e) => {
                tracing::warn!("Docker verification failed: {}, falling back to local", e);
            }
        }
    }

    // Fall back to local pytest
    run_local_pytest(workspace).await
}

/// Quick evaluation that doesn't parse detailed failures.
/// Use for frequent checks during execution.
pub async fn quick_evaluate(
    task: &TerminalBenchTask,
    workspace: &Path,
) -> Result<QuickResult> {
    let cmd = task
        .verification
        .command
        .as_deref()
        .or(task.verification.script.as_deref())
        .unwrap_or("exit 1");

    let output = Command::new("sh")
        .args(["-c", cmd])
        .current_dir(workspace)
        .output()
        .await
        .map_err(|e| HillClimberError::Workspace(format!("Quick evaluate failed: {}", e)))?;

    let stdout = String::from_utf8_lossy(&output.stdout);

    // Quick parse for pass count
    let passed_re = Regex::new(r"(\d+)\s+passed").unwrap();
    let failed_re = Regex::new(r"(\d+)\s+failed").unwrap();

    let passed = passed_re
        .captures(&stdout)
        .and_then(|c| c[1].parse::<u32>().ok())
        .unwrap_or(0);
    let failed = failed_re
        .captures(&stdout)
        .and_then(|c| c[1].parse::<u32>().ok())
        .unwrap_or(0);
    let total = passed + failed;
    let total = if total == 0 { 1 } else { total };

    let progress = if output.status.success() {
        1.0
    } else {
        passed as f64 / total as f64
    };

    Ok(QuickResult {
        passed: output.status.success(),
        progress,
        message: if output.status.success() {
            "All tests passing".to_string()
        } else {
            format!("{}/{} tests passing", passed, total)
        },
    })
}

/// Result from quick evaluation.
pub struct QuickResult {
    pub passed: bool,
    pub progress: f64,
    pub message: String,
}

// ============================================================================
// Prompt Formatting
// ============================================================================

/// Format evaluation result for injection into FM prompt.
pub fn format_for_prompt(result: &EvaluatorResult) -> String {
    let mut lines = Vec::new();

    if result.passed {
        lines.push(format!(
            "Verification: PASSED ({}/{} tests)",
            result.tests_passing, result.tests_total
        ));
    } else {
        lines.push(format!(
            "Verification: FAILED ({}/{} tests)",
            result.tests_passing, result.tests_total
        ));

        // Add first 3 failures
        for failure in result.failures.iter().take(3) {
            if failure.expected.is_some() && failure.actual.is_some() {
                lines.push(format!(
                    "  - {}: expected {}, got {}",
                    failure.test_name,
                    failure.expected.as_ref().unwrap(),
                    failure.actual.as_ref().unwrap()
                ));
            } else {
                let msg_truncated = if failure.message.len() > 100 {
                    format!("{}...", &failure.message[..100])
                } else {
                    failure.message.clone()
                };
                lines.push(format!("  - {}: {}", failure.test_name, msg_truncated));
            }
        }

        if result.failures.len() > 3 {
            lines.push(format!(
                "  ... and {} more failures",
                result.failures.len() - 3
            ));
        }

        if let Some(ref suggestion) = result.suggestion {
            lines.push(format!("Suggestion: {}", suggestion));
        }
    }

    lines.join("\n")
}

// ============================================================================
// Suggestion Generator
// ============================================================================

/// Generate a suggestion based on failures.
/// Note: This is generic - no task-specific suggestions.
pub fn generate_suggestion(failures: &[FailureDetail]) -> Option<String> {
    if failures.is_empty() {
        return None;
    }

    let first_failure = &failures[0];

    // If we have expected/actual, analyze the difference
    if let (Some(expected), Some(actual)) = (&first_failure.expected, &first_failure.actual) {
        // Try to parse as arrays and compare
        if expected.starts_with('[') && actual.starts_with('[') {
            // Count items roughly
            let expected_count = expected.matches(',').count() + 1;
            let actual_count = actual.matches(',').count() + 1;

            if expected_count > actual_count {
                return Some(format!(
                    "Missing {} matches. Check if constraints are too restrictive.",
                    expected_count - actual_count
                ));
            } else if actual_count > expected_count {
                return Some(format!(
                    "{} false positives. Check boundary conditions.",
                    actual_count - expected_count
                ));
            }
        }
    }

    // Generic suggestion
    let msg_truncated = if first_failure.message.len() > 100 {
        format!("{}...", &first_failure.message[..100])
    } else {
        first_failure.message.clone()
    };
    Some(format!("Fix: {}", msg_truncated))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_pytest_output_mixed() {
        let output = r#"
============================= test session starts ==============================
collected 10 items

tests/test_solution.py::test_basic PASSED
tests/test_solution.py::test_edge_case FAILED - AssertionError: Expected ['a', 'b'], but got ['a']

=========================== short test summary info ============================
FAILED tests/test_solution.py::test_edge_case
========================= 1 passed, 1 failed in 0.05s ==========================
"#;

        let result = parse_pytest_output(output);
        assert_eq!(result.passed, 1);
        assert_eq!(result.failed, 1);
        assert_eq!(result.total, 2);
        assert_eq!(result.failures.len(), 1);
    }

    #[test]
    fn test_parse_pytest_output_all_passed() {
        let output = "========================= 5 passed in 0.10s ==========================";
        let result = parse_pytest_output(output);
        assert_eq!(result.passed, 5);
        assert_eq!(result.failed, 0);
        assert_eq!(result.total, 5);
    }

    #[test]
    fn test_format_for_prompt() {
        let result = EvaluatorResult {
            passed: false,
            progress: 0.5,
            tests_total: 10,
            tests_passing: 5,
            failures: vec![FailureDetail {
                test_name: "test_basic".to_string(),
                line_number: None,
                expected: Some("['a', 'b']".to_string()),
                actual: Some("['a']".to_string()),
                message: "AssertionError".to_string(),
            }],
            suggestion: Some("Check edge cases".to_string()),
            raw_output: String::new(),
            duration_ms: 100,
        };

        let prompt = format_for_prompt(&result);
        assert!(prompt.contains("FAILED"));
        assert!(prompt.contains("5/10"));
        assert!(prompt.contains("test_basic"));
        assert!(prompt.contains("Suggestion:"));
    }
}
